packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -895,6 +1080,687 @@ void Add(
+        {
+          count[(CcolindAux(i) & mask) >> maskPos]++;
+        }
+      }
+      //get offset as the prefix sum for count
+      for(Ordinal i = 0; i < 16; i++)
+      {
+        offset[i + 1] = offset[i] + count[i];
+      }
+      //now for each element in [lo, hi), move it to its offset in the other buffer
+      //this branch should be ok because whichBuf is the same on all threads
+      if(whichBuf == 0)
+      {
+        //copy from *Over to *Aux
+        for(Ordinal i = lo; i < hi; i++)
+        {
+          Ordinal bucket = (CcolindOver(i) & mask) >> maskPos;
+          CcolindAux(lo + offset[bucket + 1] - count[bucket]) = CcolindOver(i);
+          CvalsAux(lo + offset[bucket + 1] - count[bucket]) = CvalsOver(i);
+          count[bucket]--;
+        }
+      }
+      else
+      {
+        //copy from *Aux to *Over
+        for(Ordinal i = lo; i < hi; i++)
+        {
+          Ordinal bucket = (CcolindAux(i) & mask) >> maskPos;
+          CcolindOver(lo + offset[bucket + 1] - count[bucket]) = CcolindAux(i);
+          CvalsOver(lo + offset[bucket + 1] - count[bucket]) = CvalsAux(i);
+          count[bucket]--;
+        }
+      }
+      whichBuf = 1 - whichBuf;
+      mask = mask << 4;
+      maskPos += 4;
+    }
+    if(whichBuf)
+    {
+      //need to deep copy from aux arrays to main
+      for(Ordinal i = lo; i < hi; i++)
+      {
+        CcolindOver(i) = CcolindAux(i);
+        CvalsOver(i) = CvalsAux(i);
+      }
+    }
+  }
+  
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const {
+    //3: Sort each row's colinds (permuting values at same time), then count unique colinds (write that to Crowptr(i))
+    //CrowptrTemp tells how many entries in each oversized row
+    Ordinal rowStart = CrowptrTemp(i);
+    Ordinal rowEnd = CrowptrTemp(i + 1);
+    Ordinal rowNum = rowEnd - rowStart;
+    radixSort2(rowStart, rowEnd);
+    //now that they are sorted, can sweep through and gather together values in same column
+    Ordinal finalPos = 0;
+    Ordinal overPos = 0;
+    while(overPos < rowNum)
+    {
+      //check if the two entries at overPos are the same col
+      if(CcolindOver(rowStart + overPos) == CcolindOver(rowStart + overPos + 1) && (overPos + 1 < rowNum))
+      {
+        //they are the same col, so combine the two entries at finalPos
+        //Then increment overPos an extra time since just processed two entries
+        CcolindOver(rowStart + finalPos) = CcolindOver(rowStart + overPos);
+        CvalsOver(rowStart + finalPos) = CvalsOver(rowStart + overPos) + CvalsOver(rowStart + overPos + 1);
+        overPos++;
+      }
+      else
+      {
+        //they are not the same col, so just copy the one entry at overPos to finalPos
+        CcolindOver(rowStart + finalPos) = CcolindOver(rowStart + overPos);
+        CvalsOver(rowStart + finalPos) = CvalsOver(rowStart + overPos);
+      }
+      overPos++;
+      finalPos++;
+    }
+    Crowcounts(i) = finalPos;
+  }
+  const RowPtrs& CrowptrTemp;
```

Comment:
See above note on Kokkos::View needing to be stored by value in functors, not by reference.